### ISSUE TITLE
[BF16] Visit all the type cast from integer to BF16 type for potential accuracy loss

### ISSUE THREAD

**Issue by jgong5** [pytorch/pytorch#89212]
## Issue description

BF16 cannot accurately represent all integer values outside the range of [-255,255]. When there is a cast from an integer value from BF16 data type, there would be potential accuracy loss. Typically, such a type cast is used in pixel index calculation in ops such as upsample, pooling etc. FP32 is preferred in these cases. See https://github.com/pytorch/pytorch/issues/88939 for a related functionality issue.

## Code example

`area_pixel_compute_scale` (https://github.com/jgong5/pytorch/blob/fc60a1865eafc985217eccc0251f82014041e6a7/aten/src/ATen/native/UpSample.h#L263) is one of the examples and used in a lot of places. There are also direct usage of `static_cast<scalar_t>` from bf16 to integer, such as https://github.com/jgong5/pytorch/blob/fc60a1865eafc985217eccc0251f82014041e6a7/aten/src/ATen/native/FractionalMaxPool2d.cpp#L139

**Comment by jgong5** [comment]
> Well, float can not accurately represent large integers either, so imo a proper fix would be to add some sort of `index = std::min(index, input_size - 1)` guard

Yes, I just planned to add the check as well. Thanks for the reminder. The reason I didn't add it the first place is that the size would be very big and the size would be too big to get allocated.

**Comment by CaoE** [comment]
There is also a lot of accumulations in bfloat16 : https://github.com/pytorch/pytorch/blob/5adc18dcbc7f66364914baaea40445a314c70ea5/aten/src/ATen/native/cpu/UpSampleMoreKernel.cpp#L53 for backward. When scale_factor is large the accumulation in bfloat16 will get wrong results.

**Comment by CaoE** [comment]
BF16 is not supported by FractionalMaxPool.

**Comment by jgong5** [comment]
It was marked as done.
